cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java

```diff
@@ -275,57 +297,129 @@ public List<TopicId> listTopics(NamespaceId namespaceId)
       throws TopicNotFoundException, IOException, UnauthorizedException {
     long start = System.currentTimeMillis();
+    CompletableFuture<Void> publishFuture = new CompletableFuture<>();
+    batch.add(new RequestFuture(request, publishFuture));
 
-    batch.add(request);
-    if (!batch.isEmpty()) {
-      int sequenceId = 0;
-      List<Mutation> batchCopy = new ArrayList<>(batch.size());
-      // We need to batch less than fetch limit since we read for publish_ts >= last_message.publish_ts
-      // see fetch for explanation of why we read for publish_ts >= last_message.publish_ts and
-      // not publish_ts > last_message.publish_ts
-      while (!batch.isEmpty()) {
-        StoreRequest headRequest = batch.poll();
-        for (byte[] payload : headRequest) {
-          //TODO: [CDAP-21094] Breakdown messages with larger payload into parts.
-          Mutation mutation = Mutation.newInsertBuilder(getTableName(headRequest.getTopicId()))
-              .set(SEQUENCE_ID_FIELD).to(sequenceId++).set(PAYLOAD_SEQUENCE_ID_FIELD).to(0)
-              .set(PUBLISH_TS_FIELD).to("spanner.commit_timestamp()").set(PAYLOAD_FIELD)
-              .to(ByteArray.copyFrom(payload)).build();
-          batchCopy.add(mutation);
-        }
+    synchronized (clientLock) {
+      writeData(start);
+    }
+    // Wait for the corresponding future to finish.
+    publishFuture.join();
+
+    //TODO: Add a RollbackDetail implementation that throws exceptions if any of the methods is called.
+    return null;
+  }
+
+  private void writeData(long start) {
+    AtomicInteger sequenceId = new AtomicInteger(0);
+    long estimatedBatchSize = 0;
+    List<Mutation> mutations = new ArrayList<>(batch.size());
+    List<CompletableFuture<Void>> publishFutures = new ArrayList<>();
+
+    // We need to batch less than fetch limit since we read for publish_ts >= last_message.publish_ts
+    // see fetch for explanation of why we read for publish_ts >= last_message.publish_ts and
+    // not publish_ts > last_message.publish_ts
+    while (!batch.isEmpty()) {
+      RequestFuture requestFuture = batch.poll();
+      StoreRequest currentRequest = requestFuture.getStoreRequest();
+
+      // Check if adding these mutations will exceed the batch size limit.
+      // If exceeding limit, publish the current batch and start a new one.
+      // Ensure that all messages within a StoreRequest are published in a single batch to maintain atomicity.
+      // If the batch publish succeeds, all messages are persisted,
+      // and the corresponding future is completed successfully.
+      // If the batch publish fails, none of the messages are persisted,
+      // and the future is completed exceptionally, allowing the caller to retry the entire StoreRequest.
+      AtomicLong currentBatchSize = new AtomicLong(0);
+      currentRequest.forEach(payload -> currentBatchSize.addAndGet(payload.length));
+      if (estimatedBatchSize + currentBatchSize.get() > MAX_BATCH_SIZE_IN_BYTES) {
```

Comment:
Throw an exception in the future of a request if its payload size is greater than MAX_BATCH_SIZE_IN_BYTES. This should be done in the publish as soon as we get the request. Not in writeData